### ISSUE TITLE
Implement syntactic sugar for symmetric tensors

### ISSUE THREAD

**Issue by egisatoshi** [egison/egison#231]
We want to desugar
```
def X{_i_j} :=
  [| [| 0, 1, 2 |],
     [| 1, 0, 3 |],
     [| 2, 3, 0 |] |]
```
to
```
def X_i_j :=
 let tmpX_i_j :=
       [| [| 0, 1, 2 |],
          [| 1, 0, 3 |],
          [| 2, 3, 0 |] |] in
   generateTensor
     (\i j -> if i > j then tmpX_j_i
              else          tmpX_i_j)
     (tensorShape tmpX_#_#)
```

**Comment by egisatoshi** [comment]
We also want to desugar
```
def X[_i_j] :=
  [| [| 0, 1, 2 |],
     [| 1, 0, 3 |],
     [| 2, 3, 0 |] |]
```
to
```
def X_i_j :=
 let tmpX_i_j :=
       [| [| 0, 1, 2 |],
          [| -1, 0, 3 |],
          [| -2, -3, 0 |] |] in
   generateTensor
     (\i j -> match (i, j) as (integer, integer) with
        | ($i, (?(\j -> i > j) & $j) -> -tmpX_j_i
        | _ tmpX_i_j)
     (tensorShape tmpX_#_#)
```

**Comment by egisatoshi** [comment]
More complex case:
```
def R{[_a_b][_c_d]} := ...
```
```
def R_a_b_c_d :=
  let tmpR_a_b_c_d := ... in
    generateTensor
      (\a b c d ->
         if      a > b, c > d, [b, a] > [d, c] then tmpR_d_c_b_a
         else if a > b, c > d, [b, a] < [d, c] then tmpR_b_a_d_c
         else if a > b, c < d, [b, a] > [c, d] then tmpR_c_d_b_a
         else if a > b, c < d, [b, a] < [c, d] then tmpR_b_a_c_d
         else if a < b, c > d, [a, b] > [d, c] then tmpR_d_c_a_b
         else if a < b, c > d, [a, b] < [d, c] then tmpR_a_b_d_c
         else if a < b, c < d, [a, b] > [c, d] then tmpR_c_d_a_b
         else if a < b, c < d, [a, b] < [c, d] then tmpR_a_b_c_d)
      (tensorShape tmpR_#_#_#_#)
```

**Comment by momohatt** [comment]
Suppose that the sub/sup of indices are consistent in one (anti-)symmetric group, then we can desugar
```
def R{[_a_b][_c_d]} := ...
```
as follows.
```
def R_a_b_c_d :=
  let tmpR_a_b_c_d := ... in
    generateTensor
      (\a b c d ->
         let (sgn1, ab) := sort [a, b]
             (sgn2, cd) := sort [c, d]
             (_, abcd)  := sort [ab, cd]
          in sgn1 * sgn2 * tmpR_(abcd_1)..._(abcd_4))
```
where `sort` is a function to sort a given collection and returns a sorted collection (tensor?) and a sign (-1 or 1). The sign is -1 if the number of swaps needed to sort the collection is even, and 1 otherwise.
```
sort [2, 1, 3] -- (-1, [1, 2, 3])
sort [3, 1, 2] -- (1, [1, 2, 3])
```

**Comment by momohatt** [comment]
TODO: Support mixing of sub/superscripts, such as
```
def R[_a~b] := ...
```

**Comment by momohatt** [comment]
Fixed in 62327bc2ee958cd4d7665419769eb13ea3c6b1b3
